firebase-appdistribution/src/main/java/com/google/firebase/appdistribution/impl/SequentialReference.java

```diff
@@ -0,0 +1,94 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.appdistribution.impl;
+
+import androidx.annotation.VisibleForTesting;
+import com.google.android.gms.tasks.Task;
+import com.google.android.gms.tasks.TaskCompletionSource;
+import com.google.firebase.appdistribution.UpdateTask;
+import com.google.firebase.concurrent.FirebaseExecutors;
+import java.util.concurrent.Executor;
+
+/**
+ * A reference to an object that uses a sequential executor to ensure non-concurrent reads and
+ * writes while preventing lock contention.
+ */
+class SequentialReference<T> {
```

Comment:
I have to admit I don't really grok this. I really hope @vkryachko does.

Or maybe there is a different paradigm that can be used? This looks like low-level infrastructure which seems to be useful in general - so I'd assume there is some established pattern to do things like this.

Reply:
I've updated this to be simpler based on our conversation today.